File: projects/test_models/Zeta/cs/Types/TestType.cs
using System;

namespace Phi
{
    /// <summary>
    /// Test type
    /// </summary>
    public sealed class TestType
    {
        #region Properties
        /// <summary>
        /// Some comment
        /// </summary>
        public Zeta.String Version { get; set; }
        public Zeta.String AField { get; set; }
        #endregion

        #region Constructors
        public TestType() { }

        public TestType(
            Zeta.String version,
            Zeta.String aField)
        {
            Version = version;
            AField = aField;
        }
        #endregion

        #region Equality
        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj)) return false;
            if (ReferenceEquals(this, obj)) return true;
            if (obj.GetType() != GetType()) return false;

            var value = obj as TestType;
            if (value == null) return false;

            return
                Version == value.Version &&
                AField == value.AField;
        }

        public static bool operator ==(TestType lhs, TestType rhs)
        {
            if (Object.ReferenceEquals(lhs, rhs))
                return true;

            return !Object.ReferenceEquals(null, lhs) && lhs.Equals(rhs);
        }

        public static bool operator !=(TestType lhs, TestType rhs)
        {
            return !(lhs == rhs);
        }

        public override int GetHashCode()
        {
            unchecked
            {
                // Choose large primes to avoid hashing collisions
                const int HashingBase = (int) 2166136261;
                const int HashingMultiplier = 16777619;

                int hash = HashingBase;
                hash = (hash * HashingMultiplier) ^ Version.GetHashCode();
                hash = (hash * HashingMultiplier) ^ AField.GetHashCode();
                return hash;
            }
        }
        #endregion
    }
}

File: projects/test_models/Zeta/cpp/src/types/TestType.cpp
#include "Phi/types/TestType.hpp"

namespace Phi {

TestType::TestType()
    : Version_(static_cast<Zeta::String>(0)),
      AField_(static_cast<Zeta::String>(0)) { }

TestType::TestType(
    const Zeta::String Version,
    const Zeta::String AField)
    : Version_(Version),
      AField_(AField) { }

void TestType::swap(TestType& other) noexcept {
    using std::swap;
    swap(Version_, other.Version_);
    swap(AField_, other.AField_);
}

bool TestType::operator==(const TestType& rhs) const {
    return Version_ == rhs.Version_ &&
        AField_ == rhs.AField_;
}

TestType& TestType::operator=(TestType other) {
    using std::swap;
    swap(*this, other);
    return *this;
}

Zeta::String TestType::Version() const {
    return Version_;
}

void TestType::Version(const Zeta::String v) {
    Version_ = v;
}

Zeta::String TestType::AField() const {
    return AField_;
}

void TestType::AField(const Zeta::String v) {
    AField_ = v;
}

}
